libs/libc/misc/lib_tempbuffer.c

```diff
@@ -63,58 +63,61 @@ static struct pathbuffer_s g_pathbuffer =
  ****************************************************************************/
 
 /****************************************************************************
- * Name: lib_get_pathbuffer
+ * Name: lib_get_tempbuffer
  *
  * Description:
- *   The lib_get_pathbuffer() function returns a pointer to a temporary
+ *   The lib_get_tempbuffer() function returns a pointer to a temporary
  *   buffer.  The buffer is allocated from a pool of pre-allocated buffers
  *   and if the pool is exhausted, a new buffer is allocated through
- *   kmm_malloc(). The size of the buffer is PATH_MAX, and must freed by
- *   calling lib_put_pathbuffer().
+ *   kmm_malloc(). The size of the buffer is nbytes, and must freed by
+ *   calling lib_put_tempbuffer().
  *
  * Returned Value:
- *   On success, lib_get_pathbuffer() returns a pointer to a temporary
+ *   On success, lib_get_tempbuffer() returns a pointer to a temporary
  *   buffer.  On failure, NULL is returned.
  *
  ****************************************************************************/
 
-FAR char *lib_get_pathbuffer(void)
+FAR char *lib_get_tempbuffer(size_t nbytes)
 {
-  for (; ; )
+  if (nbytes <= TEMP_MAX_SIZE)
```

Comment:
1. nbytes <= TEMP_MAX_SIZE: Make sure don't cross the line
2. At present, parameter passing is supported, and the code under the following app uses this interface to optimize the use of the stack. The new feature of nshlib has exceeded the call stack of 2k.

Reply:
You should call malloc in nshlib instead of making the interface so ugly

Reply:
many places (especially in nsh) need allocate temp buffer (e.g. path or line) only in function scope, both follow approach isn't perfect:

1. Allocate from stack, which already cause the default 2KB stack is unusable in many cases:
    https://github.com/apache/nuttx/pull/15173
    https://github.com/apache/nuttx/pull/14501
2. Allocate from heap, which may increase memory fragment and impact performance

so, it's reasonable to extend pathbuffer to tempbuffer, and let's caller utilize the general buffer facility when they find some function occupy too much stack space.

> You should call malloc in nshlib instead of making the interface so ugly

It's natural to add size argument to a buffer allocation function. Do you feel ugly malloc with size or not strange malloc without size?

Reply:
The implementation of pathbuffer is a memory pool with a limited length. Currently, all the kernel code is using the length of PATH_MAX. From the kernel perspective, increasing the length parameter is a strange design, because the kernel code can ensure that all locations(100%) that call pathbuffer only use PATH_MAX:

FAR char *lib_get_pathbuffer(void);
-> FAR char *lib_get_tempbuffer(size_t nbytes);   <- why?

If the customized length in the application exceeds PATH_MAX, we should manage the life cycle of this memory segment by application self. If the stack occupies too much, we can choose to define it statically or call malloc:

https://github.com/apache/nuttx-apps/blob/master/nshlib/nsh_console.h#L157

The design of the API should be more reasonable and easy to use. If 99% of the code is using PATH_MAX as the expected value, why should these codes bear the overhead of parameter passing and parameter checking? This is unreasonable

Reply:
i'm not sure how it's different from malloc either.

> Allocate from heap, which may increase memory fragment and impact performance

do you mean line buffer allocation in nsh is performance critical? really?
also, as far as the allocation lifetime is very short, fragmentation is not a big concern i suspect.

Reply:
> The implementation of pathbuffer is a memory pool with a limited length. Currently, all the kernel code is using the length of PATH_MAX. From the kernel perspective, add length parameter is a strange design, because the kernel code can ensure that all locations(100%) that call pathbuffer only use PATH_MAX:
> 
> FAR char *lib_get_pathbuffer(void); -> FAR char *lib_get_tempbuffer(size_t nbytes); <- why?
> 
> If the customized length in the application exceeds PATH_MAX, we should manage the life cycle of this memory segment by application self. If the stack occupies too much, we can choose to define it statically or call malloc:
> 
> https://github.com/apache/nuttx-apps/blob/master/nshlib/nsh_console.h#L157
> 
> The design of the API should be more reasonable and easy to use. If 99% of the code is using PATH_MAX as the expected value, why should these codes bear the overhead of parameter passing and parameter checking? This is unreasonable

if you think lib_get_pathbuffer is used frequently, we need  keep the caller as simple as possible in this case, how about:
```
#define lib_get_pathbuffer() lib_get_tempbuffer(PATH_MAX)
FAR char *lib_get_tempbuffer(size_t nbytes);
```

Reply:
No, I think kernel code shouldn't pass this parameter, instead of fooling itself with a macro